cdmtaskservice/jaws/remote.py

```diff
@@ -0,0 +1,71 @@
+"""
+Code for parsing JAWS output expected to be run at a remote location, .e.g on NERSC.
+
+In particular, non-standard lib dependency imports should be kept to a minimum and the newest
+python features should be avoided to make setup on the remote cluster simple and allow for older
+python versions.
+"""
+
+import io
+import json
+from pathlib import Path
+
+from cdmtaskservice.arg_checkers import not_falsy as _not_falsy
+
+
+def get_filenames_for_container(container_num: int) -> tuple[str, str, str]:
+    """
+    Given a container number, get the return code, stdout, and stderr filenames in that order
+    as a tuple.
+    """
+    if container_num < 0:
+        raise ValueError("container_num must be >= 0")
+    return (
+        f"container-{container_num}-rc.txt",
+        f"container-{container_num}-stdout.txt",
+        f"container-{container_num}-stderr.txt",
+    )
+
+
+def parse_errors_json(errors_json: io.BytesIO, logpath: Path) -> list[tuple[int, str | None]]:
+    """
+    Parses a JAWS errors.json file and writes the return code, stdout, and stderr files
+    to the given path, with the names of the files as
+    `container-{container number}-[rc | stdout | stderr].txt`
+    
+    Assumes there's only one container name in the json, which is the case for CTS jobs.
+    
+    Returns a list of tuples of the container return code and an error message
+    from Cromwell ordered by the container number.
+    The error message is typically only useful for someone with Cromwell familiarity.
+    """
+    # may need to use an iterative parsing strategy with `ijson` or something
+    # These files could be really big
+    _not_falsy(logpath, "logpath")
+    j = json.load(_not_falsy(errors_json, "errors_json"))
+    if not j:
+        raise ValueError("No errors in error json")
+    j = j["calls"]
+    if len(j) != 1:
+        raise ValueError("Expected only one call")
+    j = j[list(j.keys())[0]]
+    id2rc = {}
+    # Could parallelize some of this or use async if necessary... YAGNI 
+    for c in j:
+        cid = c["shardIndex"]
+        if cid in id2rc:
+            raise ValueError(f"Duplicate shardIndex: {cid}")
+        rc = c["returnCode"]
+        # never seen a structure other than this, may need changes
+        err = c["failures"][0]["message"]
+        id2rc[cid] = (rc, err)
+        rcf, sof, sef = get_filenames_for_container(cid)
+        with open(logpath / rcf, "w") as f:
+            f.write(f"{rc}\n")
+        with open(logpath / sof, "w") as f:
+            f.write(c["stdoutContents"])
+        with open(logpath / sef, "w") as f:
+            f.write(c["stderrContents"])
+    if len(id2rc.keys()) - 1 != max(id2rc.keys()):
```

Comment:
```suggestion
    if set(id2rc.keys()) != set(range(len(id2rc))):
```
This seems to align more accurately with the raised error.

Reply:
I think they check the same thing - since they keys have to be unique in the map, if the maximum key value is equal to the number of keys - 1, then there must be a key for each number from 0 to the maximum key value.

If I'm missing something LMK and I'll fix it